File: topics/examples/rabbitmq-consumer-example.go
package main

import (
	"encoding/json"
	"fmt"
	"log"

	"github.com/streadway/amqp"
)

func main() {
	// 连接RabbitMQ
	conn, err := amqp.Dial("amqp://guest:guest@localhost:5672/")
	if err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}
	defer conn.Close()

	// 创建Channel
	ch, err := conn.Channel()
	if err != nil {
		log.Fatalf("Failed to open channel: %v", err)
	}
	defer ch.Close()

	// 声明Queue
	q, err := ch.QueueDeclare(
		"order_queue", // name
		true,          // durable
		false,         // delete when unused
		false,         // exclusive
		false,         // no-wait
		nil,           // arguments
	)
	if err != nil {
		log.Fatalf("Failed to declare queue: %v", err)
	}

	// 设置预取数量（公平分发）
	err = ch.Qos(
		1,     // prefetch count
		0,     // prefetch size
		false, // global
	)
	if err != nil {
		log.Fatalf("Failed to set QoS: %v", err)
	}

	// 消费消息
	msgs, err := ch.Consume(
		q.Name, // queue
		"",     // consumer
		false,  // auto-ack (手动确认)
		false,  // exclusive
		false,  // no-local
		false,  // no-wait
		nil,    // args
	)
	if err != nil {
		log.Fatalf("Failed to register consumer: %v", err)
	}

	fmt.Println("Waiting for messages. To exit press CTRL+C")

	for msg := range msgs {
		var message map[string]interface{}
		if err := json.Unmarshal(msg.Body, &message); err != nil {
			log.Printf("Error decoding message: %v", err)
			msg.Nack(false, true) // 重新入队
			continue
		}

		fmt.Printf("Received: %v\n", message)

		// 处理消息
		// processOrder(message)

		// 确认消息
		msg.Ack(false)
	}
}

File: topics/examples/rabbitmq-consumer-example.rs
// RabbitMQ Consumer示例 (Rust)
// 依赖: amqprs = "0.10", tokio = "1.0"

use amqprs::{
    channel::{BasicAckArguments, BasicConsumeArguments, Channel},
    connection::{Connection, OpenConnectionArguments},
    consumer::AsyncConsumer,
    BasicProperties, Deliver,
};
use tokio;

struct OrderConsumer;

#[async_trait::async_trait]
impl AsyncConsumer for OrderConsumer {
    async fn consume(
        &mut self,
        channel: &Channel,
        deliver: Deliver,
        _basic_properties: BasicProperties,
        _content: Vec<u8>,
    ) {
        println!("Received: {:?}", deliver);
        
        // 处理消息
        // process_order(&content);
        
        // 确认消息
        let args = BasicAckArguments::new(deliver.delivery_tag(), false);
        channel.basic_ack(args).await.unwrap();
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // 连接RabbitMQ
    let args = OpenConnectionArguments::new("localhost", 5672, "guest", "guest");
    let connection = Connection::open(&args).await?;
    let channel = connection.open_channel(None).await?;

    // 声明Queue
    channel
        .queue_declare("order_queue", false, false, false, false, false, None)
        .await?;

    // 设置预取数量
    channel.basic_qos(1, false, false).await?;

    // 开始消费
    let args = BasicConsumeArguments::new("order_queue", "consumer-1");
    let (_ctag, _receiver) = channel.basic_consume(OrderConsumer, args).await?;

    println!("Waiting for messages. To exit press CTRL+C");
    
    // 保持运行
    tokio::signal::ctrl_c().await?;
    
    channel.close().await?;
    connection.close().await?;

    Ok(())
}
